README.md
<h1 align="center"> TheUI Sveltekit Components</h1>
<div align="center">

[![npm version](https://badgen.net/npm/v/@theui/sveltekit-components?color=red)](https://www.npmjs.com/package/@theui/sveltekit-components)
[![npm downloads](https://badgen.net/npm/dt/@theui/sveltekit-components)](https://www.npmjs.com/package/@theui/sveltekit-components)
[![license](https://badgen.net/npm/license/flowbite-svelte)](https://github.com/themesberg/flowbite-svelte/blob/main/LICENSE)
</div>

<h2 align="center">A tool of the <b>Sveltekit ecosystem</b></h2>

<div align="center">
  <img src="https://theui-beta.vercel.app/img/theui-sveltekit-components.svg" width="400px">
</div>

The **@theui/sveltekit-components** is an official component library of [**TheUI**](https://www.theui.dev) for **Sveltekit**, designed on the top of TailwindCSS. It includes the power of SvelteKit and the features of the TailwindCSS together to make an awesome tool that speedup your development process.

<br>

## **Features**

 * Accessible components
 * Dark/light mode control
 * Theming and branding just with a single line CSS
 * Highly customizable
 * RTL supports (Coming soon)
 * Fully typed

<br>

## **Components**

* General Components

  * Accordion
  * Alert
  * Badge
  * Breadcrumb
  * Button
  * Button group
  * Collapse
  * Container
  * Dark-mode switch
  * Dropdown
  * Modal
  * Navbar
  * Notification
  * Offcanvas
  * Popup (Exit and Entry popup)
  * Progress bar
  * Tabs and pills
  * Table
  * Tooltip

<br>

* Form Components
  * Form input
  * Check-box
  * Input group
  * Radio button
  * Range
  * Select
  * Textarea
  * Toggle
  </li>
</ul>

<br>

## **Installation Guide**

Install the Sveltekit Components library in your project or start a new project with the boilerplate template. Install it anyhow and build fast, deliver faster.

Follow any of the two methods to install the component library.

1. Use the boilerplate from Github.
2. Manually install the component library from scratch.

<br>

### **1. Use Boilerplate From Github**

Open your terminal and run the following command to install the boilerplate template. Replace the **project-name** with your desired project name.

```bash
# Clone the project
git clone https://github.com/theui-dev/template-kit.git app-name

# Install node modules
npm i

# Run the application
npm run dev
```

<br>

### **2.&nbsp;&nbsp; Manually Install From Scratch**

To install the component library manually, please follow the steps one by one given below.

### **2.1&nbsp;&nbsp; Install Svelte/Sveltekit**

```bash
npm create svelte@latest my-app
cd my-app

# Add Tailwind CSS to the application
npx svelte-add@latest tailwindcss

# Install node modules
npm install

# Install TheUI Sveltekit Component 
npm i @theui/sveltekit-components
```

<br>

### **2.2&nbsp;&nbsp; Add TheUI Components library**

```bash
npm i @theui/sveltekit-components
```

<br>

### **2.3&nbsp;&nbsp; Configuration**

Open TailwindCSS configuration file `tailwind.config.cjs` and update it with the following code.

```js
// 1. Update config.content
// 2. Add config.presets
const config = {
  content: [
    "./src/**/*.{html,js,svelte,ts}",
    "./node_modules/@theui/sveltekit-components/**/*.{html,js,svelte,ts}", // New line
  ],
  presets: [require("@theui/core/preset.cjs")], // New line
  darkMode: 'class',
  theme: {
    extend: {},
  },

  plugins: [],
};

module.exports = config;
```

Open `__layout.svelte` and import the core style file `style.postcss` inside the `<script>` tag. It is required for the custom color classes to work properly.

```html
<script>
  import '../../node_modules/@theui/core/style.postcss' // Add this line inside script tag
  import "../app.css"
</script>
```

And thats all. You are ready to start your awesome project now.

<br>
<br>

## **Colors and Branding**

Sveltekit has been developed based on the TailwindCSS, a powerful tool that can generate almost any CSS style and unlimited (available) colors for your design. But the question is, do you need unlimited colors for your design?

Well, there is no issue if we have a tool beside us that can generate any color we want. But the problem is the inconsistency of using it. If you have many options, it is very much possible to break the consistency, which will produce randomness in your design.

To solve this issue, the "TheUI Sveltekit" component library has introduced ten custom colors option that merges with other colors provided by TailwindCSS. You can customize your theme in just one line by changing the value of CSS variables of these custom colors only.

<br>

### **TheUI Colors & Default Values**

"TheUI Sveltekit" component library has the following CSS variables and values. All the CSS variables are prefixed with ui to avoid conflict with other vars.

```css
@layer base{
  :root {
    /* Brand colors */
    --ui-brand                    : 80 70 230;    /* #5046E6 */
    --ui-brand-active             : 67 56 202;    /* #4338CA */
    --ui-brand-secondary          : 253 230 138;  /* #FDE68A */
    --ui-brand-secondary-active   : 252 211 77;   /* #FCD34D */

    /* Text colors */
    --ui-text-on-brand            : 255 255 255;  /* #FFFFFF */
    --ui-text-on-brand-secondary  : 8 8 24;       /* #080818 */
    --ui-text-default             : 8 8 24;       /* #080818 */

    /* Background colors */
    --ui-bg-primary               : 255 255 255;  /* #FFFFFF */
    --ui-bg-secondary             : 243 244 246;  /* #F3F4F6 */
    --ui-bg-tertiary              : 229 231 235;  /* #E5E7EB */
  }

  /* To change any color for dark mode, change the color below */
  :root.dark {
    /* Text color on dark mode */
    --ui-text-default             : 229 231 235;  /* #E5E7EB */

    /* Background colors on dark mode */
    --ui-bg-primary               : 8 8 24;       /* #080818 */
    --ui-bg-secondary             : 22 22 38;     /* #161626 */
    --ui-bg-tertiary              : 32 32 48;     /* #202030 */
  }
}
```
> Brand colors are available everywhere. Text colors are not available for background and vice versa.

<br>

### **Use of Custom Colors**

You can use custom classes to use the custom colors in your design. Sveltekit Components library includes the following classes:

| CLASS | DESCRIPTION |
| --- | ----------- |
| .[...]-brand | Brand color for the component. For example, use the .bg-brand class to apply brand color on the background, .text-brand for the foreground, etc. |
| .[...]-brand-active | Slightly deeper or lighter version of the brand color. Use it to express any state like active or hover of an element. For example, use the .bg-brand-active class to apply brand color on the background, .text-brand-active for the foreground, etc. |
| .[...]-brand-secondary | Secondary brand color for the components. For example, use .bg-brand-secondary class to apply brand color on the background, .text-brand-secondary for foreground, etc. |
| .[...]-brand-secondary-active | Same as the .[...]-brand-active, but it is for secondary brand color. |
| .text-on-brand | Foreground color for the background color .bg-brand. |
| .text-on-brand-secondary | Foreground color for the background color .bg-brand-secondary. |
| .text-default | Default text/foreground color for the component library. |
| .text-primary | Default/primary background color. |
| .text-secondary | Secondary background color. |
| .text-tertiary | Tertiary background color. |

<br>

### **Color Customization**

Now, let's come to the main part of the doc - how to customize it? As you read above on this page, you can do this just with one line of code. Change the value of the corresponding CSS variable you want, and you are good to go. If you want to change other default colors, change values in the same way.

Let, your brand color is **rgb(255, 0, 0)**. Also, you want to change the default text color to **rgb(0,0,0)**. To change the value of the default color provided by the component library to your preferred one, open the `./src/app.css` or `./src/app.postcss` (the file which contains your TailwindCSS styles) and add the following line:.

```css
--ui-brand : 255 0 0;
--ui-text-default : 0 0 0;
```

Congratulations! All the components using the brand color and the default text color will change to the new color. You don't need to change anything more than that. To change other colors, follow the same process. All the colors will work in the same way.

> Instead of writing the RGB code **rgb(255, 255, 266)**, why should I write the code in this way? Because, this is how TailwindCSS works to apply opacity with any color.

<br>

## **Contributions**

Please let us know before starting work on any new features or bug. If you're interested in creating a new feature, create a feature request ticket in [Github Issues](https://github.com/theui-dev/sveltekit-components/issues). This allows for open discourse and may help prevent redudant work. This includes but is not limited to: creating new components, adding new utility features, or major alterations to existing work.

<br>

## **Copyright and license**

Code and documentation copyright 2022 the [M B Parvez](https://www.mbparvez.me), [Gosoft](https://www.gosoft.io) and [TheUI](https://www.theui.dev).

<br>

## **Copyright 2022 TheUI**

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

<br>

---

<br>

## **Special Thanks To [Gosoft.io](https://www.gosoft.io) and [BIPBY Digital](https://www.bipby.digital) for being our digital partner**

CHANGELOG.md
# Changelog

> Only the minor but important in-house versions will be added here before the release of v1.

<br>

## v0.32.0 (2022-08-15)
- Mobile nav fix

## v0.28.0 (2022-08-10)
- Navbar fixes
- Link and LinkGroup component removed

## v0.26.0 (2022-08-09)
- More dropdown bugs fixes

## v0.23.0 (2022-08-08)
- Core updated to v0.20.0
- Some packages updated to latest version
- Added support for secondary brand color
- Bugs fix
- Readme file updated

## v0.21.0 (2022-08-04)
- A huge change in Navbar components
  - Navbar now automatically responsive
  - Supports dropdown
  - New navbar components added such as, NavCollapse, NavDropdown, NavToggle
  - Classes and other style refreshed

## v0.21.0 (2022-07-31)
- Core component updated to 0.18.4
- max-width: 1320px added to CSS variable

## v0.19.0 (2022-07-29)
- Icon renamed as Svg for future compatibility

## v0.18.0 (2022-07-29)
- New components for table added: TableHead, TableBody, TableRow, TableCell
- Some minor bug fixes

## v0.17.10 (2022-07-27)
- Core updated to 0.18.3
- Navbar bug fixes
- Rewrite Popup documentation

## v0.17.7 (2022-07-27)
- position props of NavContainer changed to align
- Navbar max-w-screen-2xl changed to max-w-7xl

## v0.17.6 (2022-07-19)
- NavBrand and NavLink component issues fix

## v0.17.5 (2022-07-19)
- dropdownToggle slot renamed to label
- NavBrand component issue fix

## v0.17.4 (2022-07-19)
- NavBrand and NavLink components added
- Active tab hover color fix
- Bug fix

## v0.17.3 (2022-07-17)
- Dropdown: alignRight config property changed to alignEnd
- theui-core updated to 0.18.2

## v0.17.1 (2022-07-16)
- Alerts' bug fix
- Circle props removed from Badge
- Outline Button hover color fix
- Dropdown toggle button slot changed to dropdownToggle

## v0.17.0 (2022-07-15)
- variant props added to Alert
- All types props changed to variant 

## v0.16.1 (2022-07-11)
- isOpen props added to Accordion
- Default config file removed
- Store based config removed
- Global config disabled
- SEO components removed

## v0.15.0 (2022-06-29)
- Deprecated

## v0.14.11 (2022-06-19)
- Head component removed
- Minor bugfixes

## v0.14.8 (2022-06-18)
- Progressbar updated
- Offcanvas components' bug fix

## v0.14.4 (2022-06-17)
- Notification components modified
- Popup components modified
- Dropdown components' bug fix

## v0.14.0 (2022-06-16)
- Button slots renamed
- Dropdown component re-written
- Accordion and Collapse title content container changed to div

## v0.13.0 (2022-06-13)
- Updated to Tailwind CSS v3.1.2
- CSS Color vars renamed
- @theui/core updated to v0.15.1

## v0.12.0 (2022-06-07)
- Unused Config vars removed
- Default config values moved to under the components functions and files
- @theui/core updated to v0.12.2 with theming presets

## v0.10.10 (2022-05-29)
- scrollTheme added to NAV_CONFIG
- type props added to Button component

## v0.10.9 (2022-05-29)
- Button component removed from Toggle Theme component
- type props added to Button component
- Toggle theme error fixes

## v0.9.44 (2022-05-28)
- Icon component updated
- Added Changelog from this version

src/lib-config-able/config.ts
import { get, writable } from 'svelte/store'
import type { Writable } from 'svelte/store'
import type {
  SITE,
  CORE,
  ACCORDION_CONFIG_TYPE,
  BUTTON_CONFIG_TYPE,
  COLLAPSE_CONFIG_TYPE,
  DD_CONFIG_TYPE,
  MODAL_CONFIG_TYPE,
  NAV_CONFIG_TYPE,
  NOTIFY_CONFIG_TYPE,
  OFFCANVAS_CONFIG_TYPE,
  PROGRESS_CONFIG_TYPE,
  TAB_CONFIG_TYPE,
  TABLE_CONFIG_TYPE,
  TOOLTIP_CONFIG_TYPE,
  INPUT_CONFIG_TYPE,
  CHECKBOX_RADIO_CONFIG_TYPE,
  LABEL_STYLE_TYPE,
} from "@theui/core/types"

let SITE_CONFIG : Writable<SITE> = writable({
  siteName          : 'Sveltekit Website',
  slogan            : '',
  description       : 'This site is a demo of TheUI component library of Sveltekit',
  keywords          : null,
  logoPath          : null,
  imgPath           : null,
  siteSchema        : null,
  businessSchema    : null,
  lang              : 'en',
  fbAppID           : null,
  pageType          : null,
  twitterCardType   : null,
  twitterID         : null,
  twitterCreatorID  : null,
  video             : null,
  gtmCode           : null,
})

let THEUI_CONFIG : Writable<CORE> = writable({
  animate               : true,
  animateSpeed          : 'fast',
  isRounded             : true,
  reset                 : false,
  toggleTheme           : true,
})

let ACCORDION_CONFIG : Writable<ACCORDION_CONFIG_TYPE> = writable({
  contentActiveClass    : {
    'default' : 'border border-gray-300 dark:border-gray-700',
    'flush'   : 'border-b border-gray-300 dark:border-gray-700'
  },
  contentClass    : {
    'default' : 'border border-gray-300 dark:border-gray-700',
    'flush'   : 'border-b border-gray-300 dark:border-gray-700'
  },
  titleActiveClass      : {
    'default' : 'border border-brand bg-brand text-on-brand',
    'flush'   : 'border-b border-brand bg-brand/10 text-brand dark:text-on-brand'
  },
  titleClass            : {
    'default' : 'border border-gray-300 dark:border-gray-700',
    'flush'   : 'border-b border-gray-300 dark:border-gray-700'
  },
  type        : 'default',
})

let BUTTON_CONFIG : Writable<BUTTON_CONFIG_TYPE> = writable({
  outline       : false,
  shadow        : 'none',
})

let COLLAPSE_CONFIG : Writable<COLLAPSE_CONFIG_TYPE> = writable({})

// let DROPDOWN_ITEM_CONFIG : Writable<DD_ITEM_CONFIG_TYPE> = writable({
//   // --active            : false,
//   // activeClasses    : 'bg-tertiary',
//   // classes          : 'bg-transparent hover:bg-tertiary',
//   dividerClasses      : 'border-tertiary',
//   // headerClasses       : 'p-4 font-medium text-xs opacity-50',
//   // --iconAfterSize       : 1,
//   // --iconAfterClasses    : 'fill-current opacity-75',
//   // --iconBeforeSize      : 1,
//   // --iconBeforeClasses   : 'fill-current opacity-75',
//   // reset             : THEUI_CONFIG.reset,
// })

let DROPDOWN_CONFIG : Writable<DD_CONFIG_TYPE> = writable({
  alignRight          : false,
  backdrop            : false,
  block               : false,
  closeOnClick        : true,
  closeOnEsc          : true,
  closeOnOutsideClick : true,
  standalone          : true,
})

let MODAL_CONFIG : Writable<MODAL_CONFIG_TYPE> = writable({
  backdrop        : true,
  bodyClasses     : '',
  closeBtn        : true,
  closeOnEsc      : true,
  staticBackdrop  : false,
})

let NAV_CONFIG : Writable<NAV_CONFIG_TYPE> = writable({
  // animate               : THEUI_CONFIG.animate,
  // animateSpeed          : THEUI_CONFIG.animateSpeed,
  // classes                   : 'bg-transparent px-8',
  // height                : 'md',
  // reset                 : false,
  // rounded               : false,
  scrollAmountToShrink  : 32,
  scrollAmountToHide    : 64,
  scrollBehavior        : 'shrinkAndHide',
  scrollTheme           : 'bg-white dark:bg-secondary shadow-md',
})

let NOTIFY_CONFIG : Writable<NOTIFY_CONFIG_TYPE> = writable({})

let OFFCANVAS_CONFIG : Writable<OFFCANVAS_CONFIG_TYPE> = writable({
  backdrop: true,
  closeOnEsc: true,
  staticBackdrop: false,
})

let PROGRESS_CONFIG : Writable<PROGRESS_CONFIG_TYPE> = writable({})

let TABLE_CONFIG : Writable<TABLE_CONFIG_TYPE> = writable({
  border        : 'y',
  borderColor   : 'border-gray-200/80 dark:border-gray-800/80',
  hover         : true,
  // reset         : THEUI_CONFIG.reset,
  // size          : 'md',
  stripe        : false,
  // containerClasses  : '',
  // tableClasses      : '',
  theadClasses    : 'bg-gray-100 dark:bg-gray-800 text-left uppercase',
  // tbodyClasses      : 'text-left',
  trClasses       : '',
})

let TAB_CONFIG : Writable<TAB_CONFIG_TYPE> = writable({
  activeTabClasses  : {
    pills : 'bg-brand text-white hover:bg-brand-active',
    tabs : 'border-0 border-b-2 border-brand text-brand bg-primary',
  },
  tabClasses        : {
    pills : 'bg-secondary',
    tabs  : 'border-0 border-b-2 border-secondary bg-primary',
  }
})

let TOOLTIP_CONFIG : Writable<TOOLTIP_CONFIG_TYPE> = writable({})




// FORM CONTROLS AND DATA CONFIGS
let INPUT_CONFIG : Writable<INPUT_CONFIG_TYPE> = writable({
  // appearance      : 'default',
  // containerClasses    : 'flex flex-col',
  grow            : false,
  inputClasses      : {
    'default'   : 'bg-primary focus-within:bg-secondary border border-tertiary ring-tertiary w-full',
    'flat'      : 'border-0 border-b-2 border-tertiary outline-0 px-0 w-full',
    'material'  : '',
  },
  inputFocusClasses : {
    'default'   : 'focus-within:border-brand focus-within:ring-brand',
    'flat'      : 'border-0 border-b-2 border-brand outline-0 px-0 w-full',
    'material'  : '',
  },
  labelStyle      : null,
  // marginBottom    : 8,
  // reset           : THEUI_CONFIG.reset,
  // rounded         : THEUI_CONFIG.isRounded,
  // size            : 'md',
})

let FILE_INPUT_CONFIG : Writable<INPUT_CONFIG_TYPE> = writable({
  // appearance      : 'default',
  // containerClasses    : 'flex flex-col',
  // fileBtnClasses    : 'file:mr-4 file:border-0 file:text-sm file:font-semibold file:cursor-pointer file:bg-tertiary hover:file:bg-brand hover:file:text-on-brand dark:file:text-white text-gray-500',
  grow            : false,
  inputClasses      : {
    'default'   : 'bg-primary focus-within:bg-secondary dark:ring:border-gray-900 w-full',
    'flat'      : 'outline-0 px-0 w-full',
    'material'  : '',
  },
  inputFocusClasses : {
    'default'   : 'focus:ring-brand',
    'flat'      : 'border-0 border-b-2 outline-0 px-0 w-full',
    'material'  : '',
  },
  labelStyle      : null,
  // marginBottom    : 8,
  // reset           : THEUI_CONFIG.reset,
  // rounded         : THEUI_CONFIG.isRounded,
  // size            : 'md',
})

let LABEL_CONFIG : Writable<LABEL_STYLE_TYPE> = writable({
  // classes         : 'uppercase font-semibold',
  // color         : 'text-inherit',
  // marginBottom  : 2,
  // textSize      : 'text-xs',
})

let CHECKBOX_RADIO_CONFIG : Writable<CHECKBOX_RADIO_CONFIG_TYPE> = writable({
  // classes     : '',
  disabled  : false,
  // rounded   : THEUI_CONFIG.isRounded,
  // reset     : THEUI_CONFIG.reset,
  // size      : 'md',
})

export { SITE_CONFIG, THEUI_CONFIG, ACCORDION_CONFIG, BUTTON_CONFIG, COLLAPSE_CONFIG, // DROPDOWN_ITEM_CONFIG,
  DROPDOWN_CONFIG, MODAL_CONFIG, NAV_CONFIG, NOTIFY_CONFIG, OFFCANVAS_CONFIG, PROGRESS_CONFIG,
  TAB_CONFIG, TABLE_CONFIG, TOOLTIP_CONFIG, INPUT_CONFIG, FILE_INPUT_CONFIG, LABEL_CONFIG, CHECKBOX_RADIO_CONFIG }
